_posts/2019-08-20-service-objects.markdown

```diff
@@ -0,0 +1,97 @@
+categories: ["rails"]
+author: "cleiviane"
+---
+
+Service Objects is a controversial pattern for several different reasons: some developers like to use it, others like to use similar patterns and some think that they are just unnecessary due that fat models approach recommended to Rails apps.  
+
+Here at Ombulabs we like to use Service Objects whenever we can, we think it's a great way to keep our controllers skinny.
+
+In this post I would like to discuss my idea about service objects and why it's adopted by our team.
+
+<!--more-->
+
+## What is a Service Object?
+
+A Service is a stateless object that encapsulates a set of steps and usually has a single purpose. It's a great resource to decompose fat active record models while still maintaining the controllers thin, because we take away some validations that doesn't belong to a model or a controller.  
+
+Recently we added a new feature in our productized service FastRuby.io , to verify if the visitor is suitable for a discount or not, for that validation we need to check if the discount is inside a 24 hour window,  connect to an external API service to check if it's a valid discount code and then decide to show or hide the discounted price.
+
+This is not something we should do in a model. A model is a representation of a plain ruby object that can be instantiated. In this case we are going to instantiate any validator model, it's a good thing to avoid instantiate objects whenever it's possible.
+
+So if we can't use a model for this task, should we leave it in the controller? Let's try that:
+
+```ruby
+class PageController < ApplicationController
+
+  def index
+    @show_discount = false
+
+    if params[:discount_code].present?
+      discount = DiscountCode.find_by(discount_code: params[:discount_code])
+
+      if discount.present?
+        @show_discount = discount.created_at > 24.hours.ago
+      else
+        DiscountCode.create(discount_code: params[:discount_code])
+        @show_discount = true
+      end
+
+      if @show_discount
+        client = ExternalApi::Client.new
+        @show_discount = client.validate(discount_code)
+      end
+    end
+  end
+end
+```
+
+The problem with this approach is that we just fatten our controller, adding 13 more lines to it. And, in a better thinking, is it really the responsibility of the controller to check if the discount code is valid or not? I would say this is the kind of thing we should create a service for:
+
+```ruby
+class DiscountValidator
+  def self.validate(discount_code)
+    valid_discount = false
+
+    discount = DiscountCode.find_by(discount_code: discount_code)
+
+    if discount.present?
+      valid_discount = discount.created_at > 24.hours.ago
+    else
+      DiscountCode.create(discount_code: discount_code)
+      valid_discount = true
+    end
+
+    if valid_discount
+      client = ExternalApi::Client.new(api_key: MY_API_KEY)
+      valid_discount = client.validate(discount_code)
+    end
```

Comment:
@cleicar This is not totally accurate. We want to create a discount code record only after validating with the external API. 

That way, we only need to call the external API once per `discount_code`. If the visitor reloads the page over and over again, this code will call the API over and over again which is not ideal.

Reply:
@etagwerker I see your point. But this is how we are doing in Fastruby: https://github.com/ombulabs/fastruby.io/blob/master/app/services/roadmap_validator.rb

So we need correct the behavior in Fastruby as well, right?